cs/Serializer/VowpalWabbitDefaultMarshaller.cs

```diff
@@ -0,0 +1,358 @@
+
+            context.AppendStringExample(feature.Dictify, " {0}", value);
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="context">The marshalling context.</param>
+        /// <param name="ns">The namespace description.</param>
+        /// <param name="feature">The feature description.</param>
+        /// <param name="value">The actual feature value.</param>
+        public void MarshalFeatureStringSplit(VowpalWabbitMarshalContext context, Namespace ns, Feature feature, string value)
+        {
+            if (string.IsNullOrWhiteSpace(value))
+                return;
+
+            var words = value.Split((char[])null, StringSplitOptions.RemoveEmptyEntries);
+            foreach (var s in words)
+            {
+                var featureHash = context.VW.HashFeature(s, ns.NamespaceHash);
+                context.NamespaceBuilder.AddFeature(featureHash, 1f);
+            }
+
+            if (context.StringExample == null)
+            {
+                return;
+            }
+
+            foreach (var s in words)
+            {
+                context.AppendStringExample(feature.Dictify, " {0}", s);
+            }
+        }
+
+        /// <summary>
+        /// Transfers feature data to native space.
+        /// </summary>
+        /// <typeparam name="TKey"></typeparam>
+        /// <typeparam name="TValue"></typeparam>
+        /// <param name="context">The marshalling context.</param>
+        /// <param name="ns">The namespace description.</param>
+        /// <param name="feature">The feature description.</param>
+        /// <param name="value">The actual feature value.</param>
+        public void MarshalFeature<TKey, TValue>(VowpalWabbitMarshalContext context, Namespace ns, Feature feature, IEnumerable<KeyValuePair<TKey, TValue>> value)
+        {
+            Contract.Requires(context != null);
+            Contract.Requires(ns != null);
+            Contract.Requires(feature != null);
+
+            if (value == null)
+            {
+                return;
+            }
+
+            foreach (var kvp in value)
+            {
+                context.NamespaceBuilder.AddFeature(
+                        context.VW.HashFeature(Convert.ToString(kvp.Key), ns.NamespaceHash),
+                        (float)Convert.ToDouble(kvp.Value, CultureInfo.InvariantCulture));
+            }
+
+            if (context.StringExample == null)
+            {
+                return;
+            }
+
+            foreach (var kvp in value)
+            {
+                // TODO: not sure if negative numbers will work
+                context.AppendStringExample(
+                    feature.Dictify,
+                    " {0}:" + (context.VW.Settings.EnableStringFloatCompact ? "{1}" : "{1:E20}"),
+                    Convert.ToString(kvp.Key),
+                    (float)Convert.ToDouble(kvp.Value, CultureInfo.InvariantCulture));
```

Comment:
Convert.ToSingle? Did you also resolve your TODO? Seems like negative numbers should work here